### PR TITLE
feat: oauth2 client credentials

### PR DESCRIPTION
**Pull request by felicijus**
Provides a way to Authenticate with Oauth2Session for 0Auth2 Client Credentials (and possibility to adopt other Flows) via AuthMetadataPlugin (OAuth2MetadataPlugin) tailored to oauthlib and requests_oauthlib.

Provides posibility to adapt to non RFC 6749 compliant Authentification Server.

## Changes

- OAuth2MetadataPlugin (Class to be able Authenticate with any Authorization flow provided by oauthlib package
- OAuth2 Client Credentials (specific Implementation with oauthlib package for Client Credential Flow)
- New create channel functions for `create_camunda_cloud_channel` and 
    `create_oauth2_client_credentials_channel`
- Use given expires_in attribute in Token to reauthenticate after expiration (overridable, configurable if not given)

#### private APIs
1. OAuth2MetadataPlugin
2. Oauth2ClientCredentialsMetadataPlugin (Specializationof OAuth2MetadataPlugin)

#### public APIs
1. `create_camunda_cloud_channel`
3. `create_oauth2_client_credentials_channel`

### Deprecations *(required)*

The other way to `create_camunda_cloud_channel` (channel/camunda_cloud_channel.py) could be deleted.

### Enhancements *(optional)*

More configurable way to handle Authentification for pyzeebe grpc channel.

## Checklist

- [x] Unit tests
- [x] Documentation

## References

https://datatracker.ietf.org/doc/html/rfc6749
https://oauthlib.readthedocs.io/en/latest/oauth2/clients/client.html
https://requests-oauthlib.readthedocs.io/en/latest/oauth2_workflow.html

### CODE REVIEW

File: pyzeebe/channel/__init__.py [hunk 1]
```diff
@@ -1,3 +1,7 @@
-from pyzeebe.channel.camunda_cloud_channel import create_camunda_cloud_channel
+# from pyzeebe.channel.camunda_cloud_channel import create_camunda_cloud_channel # FIXME: could be removed
```
Comment (felicijus) [inline pyzeebe/channel/__init__.py:1]: I think this can be removed, the new implementation also supports Camunda Cloud.

File: pyzeebe/channel/oauth_channel.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+def create_oauth2_client_credentials_channel(
+    target: str,
+    client_id: str,
+    client_secret: str,
+    authorization_server: str,
+    scope: Optional[str] = None,
+    audience: Optional[str] = None,
+    expire_in: Optional[int] = None,
+    channel_options: Optional[ChannelArgumentType] = None,
```
Comment (dimastbk) [inline pyzeebe/channel/oauth_channel.py:23]: `channel_credentials` should be presented in this method for using self-signed certificates.

**Comment by dimastbk** [comment]
Hi. Do you want to replace #412?

File: pyzeebe/credentials/oauth.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    def call_credentials(self) -> grpc.CallCredentials:
+        """
+        Creates and returns the gRPC call credentials.
+
+        Returns:
+            grpc.CallCredentials: The gRPC call credentials.
+        """
+        return grpc.metadata_call_credentials(self)
+
+    def channel(self, target: str, channel_options: Optional[ChannelArgumentType] = None) -> grpc.aio.Channel:
```
[the remark this answers is not in the document]
Comment (felicijus) [reply pyzeebe/credentials/oauth.py:185]: So either moving it to the Generalisation ``OAuth2MetadataPlugin`` or don't even have it in here and only in ``pyzeebe/channel/oauth_channel.py`` ?

File: pyzeebe/channel/oauth_channel.py (hunk 2 above)
Comment (felicijus) [reply pyzeebe/channel/oauth_channel.py:23]: I will add it.

File: pyzeebe/credentials/oauth.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dimastbk) [reply pyzeebe/credentials/oauth.py:185]: Yes, I think.

**Comment by felicijus** [comment]
I think the changes are now all implemented. I needed to change the commit author therefore the force push.

File: pyzeebe/channel/__init__.py [hunk 4]
```diff
@@ -1,3 +1,10 @@
-from pyzeebe.channel.camunda_cloud_channel import create_camunda_cloud_channel
+from pyzeebe.channel.camunda_cloud_channel import (
+    create_camunda_cloud_channel,  # FIXME: could be removed
```
Comment (felicijus) [inline pyzeebe/channel/__init__.py:2]: Left it in so the "old" way for channel creation as default. 
For the usage of the new oauth channels they would have to be imported with e.g. `from pyzeebe.channel.oauth_channel import create_oauth2_client_credentials_channel`.

If we want to have oauth as the default, we could uncomment the lines below and phase out the old way. After that we could delete the "old" way altogether.

File: pyzeebe/credentials/oauth.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
+import json
+import logging
+import time
+import timeit
+from collections.abc import Callable
```
Comment (felicijus) [inline pyzeebe/credentials/oauth.py:5]: Will change this to `from collections.abc import Callable`.
https://docs.python.org/3/library/typing.html#callable

**Comment by felicijus** [comment]
How should we handle the mypy errors @dimastbk ?

I would suggest doing it with inline ignore statement `# type: ignore`.
````shell
pyzeebe/credentials/oauth.py:10: error: Skipping analyzing "grpc._auth": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pyzeebe/credentials/oauth.py:17: error: Class cannot subclass "AuthMetadataPlugin" (has type "Any")  [misc]
pyzeebe/channel/oauth_channel.py:[5](https://github.com/camunda-community-hub/pyzeebe/actions/runs/10503175124/job/29127825861?pr=468#step:6:6): error: Skipping analyzing "grpc.aio._typing": module is installed, but missing library stubs or py.typed marker  [import-untyped]
pyzeebe/channel/oauth_channel.py:5: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
Found 3 errors in 2 files (checked 52 source files)
Error: Process completed with exit code 1.
````

File: pyzeebe/channel/oauth_channel.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
+from functools import partial
+from typing import Optional
+
+import grpc
+from grpc.aio._typing import ChannelArgumentType  # type: ignore[import-untyped]
```
Comment (dimastbk) [inline pyzeebe/channel/oauth_channel.py:5]: 1) We should not import from private modules of external libs.
2) It does not make sense to import smth for typing purposes and marking it as untyped.
3) ChannelArgumentType (tuple) != channel_options from other channel constructors.

We should use `Optional[Dict[str, Any]]` and `get_channel_options`.

File: pyzeebe/channel/oauth_channel.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+
+def create_oauth2_client_credentials_channel(
+    target: str,
+    client_id: str,
+    client_secret: str,
+    authorization_server: str,
+    scope: Optional[str] = None,
+    audience: Optional[str] = None,
+    channel_credentials: grpc.ChannelCredentials = grpc.ssl_channel_credentials(),
```
Comment (dimastbk) [inline pyzeebe/channel/oauth_channel.py:17]:
```suggestion
    channel_credentials: Optional[grpc.ChannelCredentials] = None,
```

File: pyzeebe/credentials/oauth.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
+import json
+import logging
+import time
+import timeit
+from functools import partial
+from typing import Any, Callable, Optional
+
+import grpc
+import requests
+from grpc._auth import _sign_request  # type: ignore[import-untyped]
```
Comment (dimastbk) [inline pyzeebe/credentials/oauth.py:10]: Import from private module

File: pyzeebe/channel/oauth_channel.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+from typing import Optional
+
+import grpc
+from grpc.aio._typing import ChannelArgumentType  # type: ignore[import-untyped]
+
+from pyzeebe.credentials.oauth import Oauth2ClientCredentialsMetadataPlugin
+
+
+def create_oauth2_client_credentials_channel(
+    target: str,
```
Comment (dimastbk) [inline pyzeebe/channel/oauth_channel.py:11]:
```suggestion
    hostname: Optional[str] = None,
    port: Optional[int] = None,
```
As other constructors

File: pyzeebe/channel/__init__.py (hunk 4 above)
Comment (dimastbk) [reply pyzeebe/channel/__init__.py:2]: I'm not familiar with Camunda Cloud. Let us choose only one way to create a channel for it.

Comment (felicijus) [reply pyzeebe/channel/__init__.py:2]: The new way would be more consistent.

File: pyzeebe/channel/oauth_channel.py (hunk 9 above)
Comment (felicijus) [reply pyzeebe/channel/oauth_channel.py:11]: We can do it like this, but: 
1. I don't think there is much of a benefit, users could just use `target=f"{hostname}:{port}`
2. Camunda Java Client only has one variable.
https://docs.camunda.io/docs/apis-tools/java-client/#bootstrapping

File: pyzeebe/channel/oauth_channel.py (hunk 7 above)
Comment (felicijus) [reply pyzeebe/channel/oauth_channel.py:17]: To be honest I don't really know what's better.
My reasoning was that with the default value the user gets nudged in the right direction regarding which function to use.

File: pyzeebe/credentials/oauth.py (hunk 8 above)
Comment (felicijus) [reply pyzeebe/credentials/oauth.py:10]: ✅ Will change it.

Comment (felicijus) [reply pyzeebe/credentials/oauth.py]: ✅ Done

File: pyzeebe/channel/oauth_channel.py (hunk 6 above)
Comment (felicijus) [reply pyzeebe/channel/oauth_channel.py]: Commited the necessary changes to be compliant with linked grpc documentation.

File: pyzeebe/channel/camunda_cloud_channel.py [hunk 10]
```diff
@@ -1,23 +1,31 @@
...
 
 from pyzeebe.channel.channel_options import get_channel_options
+from pyzeebe.credentials.typing import ChannelArgumentType
 from pyzeebe.errors import (
     InvalidCamundaCloudCredentialsError,
     InvalidOAuthCredentialsError,
 )
 
 
+@deprecation.deprecated(
```
Comment (dimastbk) [inline pyzeebe/channel/camunda_cloud_channel.py:17]: https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.deprecated

File: pyzeebe/channel/channel_options.py [hunk 11]
```diff
@@ -8,27 +8,34 @@
...
+        existing = set()
+        _options = []
+
+        for a, b in (*options, *GRPC_CHANNEL_OPTIONS_DEFAULT):
+            if a not in existing:  # NOTE: Remove duplicates, fist one wins
+                existing.add(a)
+                _options.append((a, b))
+
+        return tuple(_options)  # (*options, GRPC_CHANNEL_OPTIONS)
+    return GRPC_CHANNEL_OPTIONS_DEFAULT
```
Comment (dimastbk) [inline pyzeebe/channel/channel_options.py:41]: I think it (changing options from dict to tuple) should be in another PR.

File: pyzeebe/channel/oauth_channel.py (hunk 9 above)
Comment (dimastbk) [reply pyzeebe/channel/oauth_channel.py:12]: Yes, but now we use host and port as two args in other methods. I think, if you want to merge it to one `target`, it should be in another PR.

File: docs/credentials.rst [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+   )
+   call_credentials: grpc.CallCredentials = grpc.metadata_call_credentials(oauth2_client_credentials)
+   channel_credentials: grpc.ChannelCredentials = grpc.ssl_channel_credentials(
+      certificate_chain=None, private_key=None, root_certificates=None
+   )
+   composite_credentials: grpc.ChannelCredentials = grpc.composite_channel_credentials(
+      channel_credentials, call_credentials
+   )
+   options: ChannelArgumentType = (("grpc.so_reuseport", 0),)
+   channel: grpc.aio.Channel = grpc.aio.secure_channel(
```
Comment (dimastbk) [inline docs/credentials.rst:32]: `pyzeebe.create_secure_channel`

File: pyzeebe/credentials/oauth.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+        """
+        Sets the expiration time for the token if it is not provided in the response.
+
+        Args:
+            r (requests.Response): The response object containing the token.
+
+        Returns:
+            requests.Response: The modified response object with the updated token.
+        """
+        token = json.loads(r.text)
```
Comment (dimastbk) [inline pyzeebe/credentials/oauth.py:124]:
```suggestion
        token = r.json()
```

**Review by dimastbk — COMMENTED** [review]
Sorry, I've been on vacation for the last two weeks.

File: pyzeebe/channel/channel_options.py (hunk 11 above)
Comment (felicijus) [reply pyzeebe/channel/channel_options.py:41]: Ok, i will copy this in another PR then...

File: docs/credentials.rst (hunk 12 above)
Comment (felicijus) [reply docs/credentials.rst:32]: This just documents the most explicit way of using Oauth2ClientCredentialsMetadataPlugin, normally these classes are not used directly anyways, so I think this should be fine.

File: pyzeebe/channel/camunda_cloud_channel.py (hunk 10 above)
Comment (felicijus) [reply pyzeebe/channel/camunda_cloud_channel.py]: ✅ Used [typing_extensions.deprecated](https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.deprecated) and removed the deprecation package.

File: pyzeebe/credentials/oauth.py (hunk 13 above)
Comment (felicijus) [reply pyzeebe/credentials/oauth.py]: ✅ Changed it.

File: pyzeebe/channel/oauth_channel.py (hunk 9 above)
Comment (felicijus) [reply pyzeebe/channel/oauth_channel.py:12]: Ok, i will soon provide another PR... 🤖

File: docs/credentials.rst (hunk 12 above)
Comment (dimastbk) [reply docs/credentials.rst:32]: Maybe add a note about default options? (keep-alive)

**Comment by felicijus** [comment]
Got some time now.
I will create a pull request regarding the grpcAddress (java zeebe) or target.
So naming could be:

1. grpc_address (favourite)
2. target

And then finish up this pull request.

Coverage should be fine, i think ✅

**Review by dimastbk — APPROVED** [review]

File: docs/channels.rst [hunk 14]
```diff
@@ -3,6 +3,22 @@ Channels
...
+.. seealso::
+    
+    `Python Channel Options <https://grpc.github.io/grpc/python/glossary.html#term-channel_arguments>`_
+      Documentation of the available Python `grpc.aio.Channel` `options` (channel_arguments).
+
+
+.. note::
+    
+    By default, channel_options is defined so that the grpc.keepalive_time_ms option is always set to 45_000 (45 seconds).
+    Reference Camunda Spring Zeebe SDK with predefined `keep-alive <https://docs.camunda.io/docs/apis-tools/spring-zeebe-sdk/configuration/#keep-alive>`_ attribute.
```
[the remark this answers is not in the document]
Comment (felicijus) [reply docs/channels.rst]: ✅ Done.

**Comment by dimastbk** [comment]
Thanks
